net: lwm2m: support NET_SOCKETS_OFFLOAD in peer parsing

junqingzou opened the pull request:
Required when offloaded socket is used
Otherwise getaddrinfo() retuns DNS_EAI_ADDRFAMILY
Desired address family is set to AF_INET

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/18765

junqingzou commented:
@mike-scott it's needed for Nordic nRF91 as we support NET_SOCKETS_OFFLOAD only

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4202,13 +4202,15 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 int lwm2m_parse_peerinfo(char *url, struct sockaddr *addr, bool *use_dtls)
```
Can you elaborate in commit message, why something like this is required then offloading is used, it is not obvious.

junqingzou replied:
CONFIG_NET_IPV6 and CONFIG_NET_IPV4 initialized hints.ai_family. If CONFIG_NET_SOCKETS_OFFLOAD does not do the same, getaddrinfo() (implemented by Nordic offload sockets) returns error -9:

> [00:00:11.420,410] <inf> net_lwm2m_engine: Parse url: coap://leshan.eclipseprojects.io
> [00:00:11.420,471] <err> net_lwm2m_engine: Unable to resolve address (-9)

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4202,13 +4202,15 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 {
 	struct http_parser_url parser;
-#if defined(CONFIG_DNS_RESOLVER)
+#if defined(CONFIG_DNS_RESOLVER) || defined(CONFIG_NET_SOCKETS_OFFLOAD)
```
This does not look right, the DNS resolving is now "done" even if is not enabled in config.

junqingzou replied:
Thanks for the comment. I added to define CONFIG_DNS_RESOLVER in my application and confirmed no issue using either hostname or IP address for LwM2M server. Thus I updated my commit with a new push.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4202,13 +4202,15 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 	u16_t off, len;
 	u8_t tmp;
 
+	LOG_INF("Parse url: %s", log_strdup(url));
```
Please use LOG_DBG() here, the info level should be used only when really needed and this does not look like that here.

junqingzou replied:
Understood, will do. This print is added only because I need to change to use different LwM2M servers, thus to leave a record in the log.

mike-scott reviewed (approved):
Looks good from my perspective.

mike-scott reviewed (changes requested):
I changed my mind on the approval.

@junqingzou This needs a more descriptive commit message explaining why the need for the change.  Something like:

The LwM2M implementation for DNS resolving has checks which configure hints based on whether IPv4 or IPv6 are enabled.   Neither of these need to be enabled when using NET_SOCKETS_OFFLOAD and this causes an error to be returned to due to hints.ai_family not being set.  Also, the offload API needs to know when to fee the allocated `struct addrinfo` instead of calling free() generically, let's use the freeaddrinfo() API for sockets which will call into the offload API if needed.

junqingzou committed:
net: lwm2m: support NET_SOCKETS_OFFLOAD in peer parsing

The LwM2M implementation for DNS resolving has checks which
configure hints based on whether IPv4 or IPv6 are enabled.
Neither of them need enabled if using NET_SOCKETS_OFFLOAD,
which then causes an error to be returned to due to
"hints.ai_family" not being set.

Also the offload API need to know when to free the allocated
"struct addrinfo" instead of calling free() generically,
thus let's use the freeaddrinfo() API for sockets which will
call into the offload API if needed.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/18765

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

junqingzou commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4282,7 +4287,11 @@ int lwm2m_parse_peerinfo(char *url, struct sockaddr *addr, bool *use_dtls)
 
 		memcpy(addr, res->ai_addr, sizeof(*addr));
 		addr->sa_family = res->ai_family;
+#if defined(CONFIG_NET_SOCKETS_OFFLOAD)
```
@mike-scott thanks for the comment. I just updated the code and commit message.

mike-scott reviewed (approved):
LGTM.  Thank you for updating.

mike-scott commented:
@jukkar @ioannisg Now that I think about this, it's a medium class bug that LwM2M subsystem doesn't support DNS when using the offload sockets API (it works for non-offload use cases).  This is a low risk patch to take into v2.0 (considering the other patches we've taken recently) and would help not only the modem drivers, but also other offloaded devices such as Wifi, etc.

I opened the following issue for that purpose: https://github.com/zephyrproject-rtos/zephyr/issues/18765

mike-scott commented:
@junqingzou  You will need to update your commit message with a "Fixes:" entry near the bottom (above your Signed-off-by).  It should look like this:
```
The LwM2M implementation for DNS resolving has checks which
configure hints based on whether IPv4 or IPv6 are enabled.
Neither of them need enabled if using NET_SOCKETS_OFFLOAD,
which then causes an error to be returned to due to
"hints.ai_family" not being set.

Also the offload API need to know when to free the allocated
"struct addrinfo" instead of calling free() generically,
thus let's use the freeaddrinfo() API for sockets which will
call into the offload API if needed.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/18765

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>
```

junqingzou commented:
@mike-scott commit message updated

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4268,6 +4270,9 @@ int lwm2m_parse_peerinfo(char *url, struct sockaddr *addr, bool *use_dtls)
 		hints.ai_family = AF_INET6;
 #elif defined(CONFIG_NET_IPV4)
 		hints.ai_family = AF_INET;
+#elif defined(CONFIG_NET_SOCKETS_OFFLOAD)
```
IMHO this is conceptually wrong thing to do. Sockets offloading has nothing to do with DNS resolving so we should not have the check like this here. I think we need to think more about this. Is this fix really needed for 2.0?

junqingzou replied:
@jukkar Please consider the situation that neither CONFIG_NET_IPV4 nor CONFIG_NET_IPV6 are defined when CONFIG_NET_SOCKETS_OFFLOAD is used and the IP stack is not in Zephyr.

junqingzou replied:
@jukkar what if we change the flag to CONFIG_NET_OFFLOAD? It's the same level of configuration as CONFIG_NET_IPV4 and CONFIG_NET_IPV6, as seen in net_context.

jukkar replied:
Changing to CONFIG_NET_OFFLOAD is no use here. The point is that we do here "magic" by checking CONFIG_NET_SOCKETS_OFFLOAD and then doing IPv4 DNS query. One option might be to create a new Kconfig option just for LWM2M, depend that on DNS and IPv4, and use that option here.

jukkar replied:
@mike-scott Can we postpone this to 2.1?

jukkar replied:
If this is really needed for 2.0, then we can do it like this. But this would need to be "fixed" properly to 2.1, perhaps #18153 helps here.

jukkar replied:
@mike-scott I am fine with that.

junqingzou replied:
@mike-scott default handling in Nordic offload sockets implementation would work but it's just a workaround. Furthermore, AF_UNSPEC does mean something and we can't overwrite it with AF_INET in all cases. Hopefully we could use current commit and wait for new solution based on [#18153 ](https://github.com/zephyrproject-rtos/zephyr/pull/18153) after 2.1

mike-scott replied:
@junqingzou Ok, let's use this for now.  I will also have more modems using offloaded DNS in Zephyr 2.1 so we will have more information to examine then.

mike-scott reviewed (changes requested):
Quick Discussion.  Don't Merge Just yet.

ioannisg commented:
> 
> 
> Quick Discussion. Don't Merge Just yet.

Not to be merged @mike-scott ?

jukkar commented:
> Not to be merged @mike-scott ?

This can be merged, @mike-scott  just had some questions about AF_UNSPEC but that one is cleared now.
